api/app/clients/sms/sap.py

```diff
@@ -110,15 +110,26 @@ def send_sms(self, to, content, reference, sender=None):
         }
 
         try:
-            resp = sms_api.send_sms_using_post(body=body)
-            order = resp.livelink_order_ids[0]
-            self.logger.info(f"SAP send SMS request for {reference} succeeded: {order.livelink_order_id[0]}")
-
-            # Avoid circular imports by importing this file later.
-            from app.models import (
-                NOTIFICATION_SENDING
-            )
-            return order.livelink_order_id[0], NOTIFICATION_SENDING
+            return self.post_sms_request(sms_api=sms_api, body=body, reference=reference)
+        except ApiException as e:
+            if e.status == 401:
+                key = 'auth'
+                auth_api = AuthorizationApi(ApiClient(Configuration(
+                    username=self._client_id,
+                    password=self._client_secret,
+                )))
+
+                resp = auth_api.token_authorization_using_post1('client_credentials')
+                access_token = resp.access_token
+                auth_cache.set(key, access_token, timeout=30 * 60)
```

Comment:
we already have logic to get this token  -- we should just lean on that existing logic

also we should remove this nested exception handling logic. we can do this by clearing the cache and re-calling this method with some flag to indicate whether we are retrying or not

Reply:
Earlier i was not sure whether to call the same method again and not to mix up with retry functionality.  Thanks for your suggestion to use a flag to differentiate the workflow.